source/task_scheduler.c

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/common/task_scheduler.h>
+#include <aws/common/thread.h>
+#include <aws/common/priority_queue.h>
+
+static const size_t DEFAULT_QUEUE_SIZE = 7;
+
+struct task_container {
+    uint64_t timestamp;
+    struct aws_task task;
+};
+
+static int compare_timestamps(const void *a, const void *b) {
+    uint64_t a_time = ((struct task_container*)a)->timestamp;
+    uint64_t b_time = ((struct task_container*)b)->timestamp;
+    return a_time > b_time; // min-heap
+}
+
+int aws_task_scheduler_init_default(struct aws_task_scheduler *scheduler, aws_task_scheduler_clock clock, 
+        struct aws_allocator *alloc) {
+    scheduler->alloc = alloc;
+    scheduler->clock = clock;
+    scheduler->owning_thread = aws_thread_current_thread_id();
+    return aws_priority_queue_dynamic_init(&scheduler->queue, alloc, DEFAULT_QUEUE_SIZE, sizeof(struct task_container), 
+            &compare_timestamps);
+}
+
+void aws_task_scheduler_init_with_queue(struct aws_task_scheduler *scheduler,
+        struct aws_priority_queue *queue, aws_task_scheduler_clock clock) {
+
+    scheduler->alloc = NULL;
+    scheduler->clock = clock;
+/* I feel icky about copying the queue. If we don't copy it, then who owns the queue? this is making me question this whole API. Need feedback */
```

Comment:
This is one reason why putting complex datastructures in caller-allocated fields can be troublesome. How do you safely move them? Do they have internal pointers-to-self that would make a simple copy unsafe? You don't know, unless you make assumptions of the implementation of the datastructure in question.

It's much cleaner to just always allocate these datastructures on the heap; platforms with difficulty allocating on the heap can use a custom allocator as an escape hatch here.

Reply:
I honestly don't know what's complicated about taking a reference to the queue and not copying it. If the user passed, it the user is responsible for cleaning it up at the appropriate time.